pkg/logql/log/parser.go

```diff
@@ -53,139 +52,110 @@ func (j *JSONParser) Process(line []byte, lbs *LabelsBuilder) ([]byte, bool) {
 	if lbs.ParserLabelHints().NoLabels() {
 		return line, true
 	}
-	it := jsoniter.ConfigFastest.BorrowIterator(line)
-	defer jsoniter.ConfigFastest.ReturnIterator(it)
-
-	// reset the state.
-	j.buf = j.buf[:0]
 	j.lbs = lbs
+	requiredLabels := j.lbs.ParserLabelHints().RequiredLabels()
+	if len(requiredLabels) > 0 {
+		err := j.parseJSONKeyVal(line, requiredLabels)
+		if err != nil {
+			lbs.SetErr(errJSON)
+		}
+		return line, true
+	}
 
-	if err := j.readObject(it); err != nil {
+	err := j.handleJson("", line)
+	if err != nil {
 		lbs.SetErr(errJSON)
 		return line, true
 	}
+
 	return line, true
 }
 
-func (j *JSONParser) readObject(it *jsoniter.Iterator) error {
-	// we only care about object and values.
-	if nextType := it.WhatIsNext(); nextType != jsoniter.ObjectValue {
-		return errUnexpectedJSONObject
-	}
-	_ = it.ReadMapCB(j.parseMap(""))
-	if it.Error != nil && it.Error != io.EOF {
-		return it.Error
+func (j *JSONParser) handleJson(preKey string, line []byte) error {
+	err := jsonparser.ObjectEach(line, func(key []byte, val []byte, valueType jsonparser.ValueType, offset int) error {
+		jsonKey := preKey + sanitizeLabelKey(string(key), true)
```

Comment:
This line here, specially string(key) makes the whole PR back to where it was before.

I suggest you take a look at difference in memory allocations, we're trying to lower them and I don't think it's happening.

Use benstats to compare results.

Reply:
I suggest you also look at benchmark profile, go test -bench -memprofile mem.out to figure where the work needs to be done.

Again the whole point is to avoid allocation as much as possible.

Reply:
From a quick look, it seems that we need to create a pool for the slice of bytes that will contains the potential keys. Then try to work with slice of bytes only, avoiding the string(key) as much as possible.